src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -27,6 +27,16 @@ internal JsonElement(JsonDocument parent, int idx)
             _idx = idx;
         }
 
+        // Currently used only as an optimization by the serializer, which does not want to
+        // return elements that are based on the <see cref="JsonDocument.Dispose"/> pattern.
+        internal static JsonElement ParseValue(ref Utf8JsonReader reader)
+        {
+            bool ret = JsonDocument.TryParseValue(ref reader, out JsonDocument? document, shouldThrow: true, useArrayPools: false);
+
+            Debug.Assert(ret, "Parse returned false with shouldThrow: true.");
+            return document!.RootElement;
```

Comment:
Do we need the null-bang? If `ret` is true, then document won't be null, right, `[NotNullWhen(true)]`?

Reply:
No check for the return value is included here, if it returns `false` then document could be null.

Reply:
But we have `Debug.Assert(ret)`. Does that not count?

cc @safern

Reply:
Yeah that should count.

Reply:
```suggestion
            return document.RootElement;
```

Reply:
It needs the bang:

```
Error	CS8602	Dereference of a possibly null reference.	System.Text.Json (net5.0), System.Text.Json (netcoreapp3.0)	D:\git\runtime6\src\libraries\System.Text.Json\src\System\Text\Json\Document\JsonElement.cs	37
```

Reply:
@safern, @stephentoub - what are we missing here?

Reply:
> what are we missing here?

The compiler doesn't track the relationship of multiple values; once you store the return value of TryParseValue into the bool, it loses the connection between the meaning of the bool and the document being non-null.  If you want to avoid the !, you can instead assert that document is not null, e.g.
```C#
JsonDocument.TryParseValue(ref reader, out JsonDocument? document, shouldThrow: true, useArrayPools: false);
Debug.Assert(document != null, "null document returned with shouldThrow: true.");
return document.RootElement;
```

Reply:
Hmm. That approach could work since I tend to be weary of any `null-forgiving` uses.

That makes me think we should add a `JsonDocument.ParseValue` internal method that throws an exception (equivalent to TryParseValue with shouldThrow = true), as an alternative to `TryParseValue`. I feel bad suggesting that we ignore the return value of a `TryX` method.

Barring that, I like your suggestion, @stephentoub as an alternative :)

Reply:
Ha, when I read your suggestion somehow I thought we were asserting in the property itself. Sorry for the miss-direction here. 

But yeah, `Debug.Assert(document != null....)` would be the way to go.

Reply:
> That makes me think we should add a JsonDocument.ParseValue internal method 

If there is a good pattern, I'm open, but I don't see one. The internal JsonDocument.TryParseValue method that is being called either throws or returns false on error depending on the `shouldThrow` parameter. This PR didn't change that. Changing to two methods would cause duplication of code.